### ISSUE TITLE
Code not working/Incomplete code

### ISSUE THREAD

**Issue by sri-harsh** [muhammedtalo/COVID-19#3]
Code not working/Incomplete code for binary classes for covid detection.

The kfold is not done in the code and moreover not able to run the code despite adding.

Changing the code to bug free and complete version would be helpful!

**Comment by muhammedtalo** [comment]
Please take a look at Issue 2 (3 classes). "yaelbeng" user is able to run the code too.

**Comment by sri-harsh** [comment]
Did not work for me, can you please help. Encountering so many errors.

**Comment by muhammedtalo** [comment]
Please make sure that you are using  "fastai" (v1) and use the code as it is given.

**Comment by muhammedtalo** [comment]
After that if it still does not work share the errors with your code.

**Comment by sri-harsh** [comment]
The error:
![image](https://user-images.githubusercontent.com/44131849/105396242-aa06c000-5c45-11eb-9195-46568357d6aa.png)
Dataset structure in the folder
![image](https://user-images.githubusercontent.com/44131849/105396286-b854dc00-5c45-11eb-83ef-7b6ccd976f4b.png)
![image](https://user-images.githubusercontent.com/44131849/105396313-c1de4400-5c45-11eb-81e0-e3ba9d01f2ad.png)

**Comment by sri-harsh** [comment]
> After that if it still does not work share the errors with your code.

Yeah, using the right version of fast.ai, the problem is with the dataset I presume. Please do guide me.

**Comment by muhammedtalo** [comment]
I would like to see inside your valid folder

**Comment by sri-harsh** [comment]
This is the valid folder:
![image](https://user-images.githubusercontent.com/44131849/105397388-00283300-5c47-11eb-9dd1-bc6808b41bd3.png)

**Comment by muhammedtalo** [comment]
Make sure that inside **Train and Valid Folder you have all three class folders**. 
and try following

data = (ImageItemList.from_folder(path, train='Train', valid ='Valid) .random_split_by_pct() .label_from_folder() .transform(tfms, size=224) .databunch())

**Comment by sri-harsh** [comment]
Not able to follow can you please explain once again, I am working on the binary class one right now not the three classes one.

**Comment by muhammedtalo** [comment]
Do you have Covid-19, Pneumonia, and No_finding class folders in **both** your Train and Valid folders ?

**Comment by sri-harsh** [comment]
You mean along with the x-ray images in train and valid folders should we have those folders as well?

**Comment by sri-harsh** [comment]
Can you share the data folder over here, how exactly it has to be? For 2 classes(binary)

https://drive.google.com/drive/folders/1efrP6t6TZBkH-bYwkxYmdZl5pQLJ3sU_?usp=sharing

**Comment by muhammedtalo** [comment]
Your folder structure should be 

X-ray Imagedataset 

       Train
            Covid-19/images
            Pneumonia/images
            No_finding/images
    
Basically, place all three class folders inside Train folder

**Comment by sri-harsh** [comment]
Still the same error

**Comment by muhammedtalo** [comment]
There is no Valid set. You should have just the Train folder. 

path='..... Dataset'/**'Train'**
data = (ImageItemList.from_folder(path)
        .random_split_by_pct()
        .label_from_folder()
        .transform(tfms, size=224)
        .databunch())

**Comment by muhammedtalo** [comment]
Is it work? if not, show the error.

**Comment by sri-harsh** [comment]
Below error is coming
![image](https://user-images.githubusercontent.com/44131849/105400177-9a3daa80-5c4a-11eb-8b8f-9ad4c79b3b19.png)

**Comment by muhammedtalo** [comment]
Try this one
path='..... Dataset'  **(without Train)**

data = ImageDataBunch.from_folder(path, train="Train", valid_pct =0.2,
        ds_tfms=get_transforms(), size=(256,256), bs=32, num_workers=4).normalize()

**Comment by muhammedtalo** [comment]
Did it work?

**Comment by sri-harsh** [comment]
No :(

![image](https://user-images.githubusercontent.com/44131849/105401499-564ba500-5c4c-11eb-9226-6ce165a04943.png)

**Comment by muhammedtalo** [comment]
Sorry, I could not understand the error. You may google it. 
Also take a look at https://forums.fast.ai/t/not-working-data-imagedatabunch-from-folder-path-ds-tfms-tfms-size-224/30510

**Comment by sri-harsh** [comment]
Can you share the most recent working copy of the code you have please?

**Comment by muhammedtalo** [comment]
The code we have shared is the most recent one. See it worked over there. All the epochs and results

**Comment by sri-harsh** [comment]
Ok, thank you for your time. Means a lot!

**Comment by muhammedtalo** [comment]
You can also try fastai-v2 which is current version. But you should update the code (https://docs.fast.ai/tutorial.vision.html)
Please let me know if that works?

**Comment by sri-harsh** [comment]
Yeah, I will check it out. Thanks for the suggestion!

**Comment by muhammedtalo** [comment]
In v2 colab

!pip install -Uqq fastbook
import fastbook
fastbook.setup_book()


from fastbook import *

from fastai.vision.all import *
path = Path('/content/gdrive/My Drive/.../')
train_fnames = get_image_files(path/"Train")


dblock = DataBlock(blocks=(ImageBlock(cls=PILImage), CategoryBlock),
                   splitter=RandomSplitter(valid_pct=0.2, seed=23),
                   get_y=parent_label,
                   item_tfms=Resize(512, method="squish"),
                   batch_tfms=None,
                   )

dls = dblock.dataloaders(train_fnames)

train_df = len(train_fnames)
print(f"There are {train_df} images")

# print class names, number of class, length of train dataset, and length of valid dataset
dls.c, len(dls.train_ds), len(dls.valid_ds)

dls.show_batch()

....

**Comment by muhammedtalo** [comment]
Please let me know it works. I would like to further help if needed.

**Comment by sri-harsh** [comment]
Ok, I am trying it out. Will let you know.

**Comment by sri-harsh** [comment]
Getting the below error

![image](https://user-images.githubusercontent.com/44131849/105404607-2acab980-5c50-11eb-8abd-ea0f47817bd5.png)

**Comment by muhammedtalo** [comment]
path should contain your file directory
For example, 
path = Path('/content/gdrive/My Drive/X-ray Image Dataset/')

**Comment by muhammedtalo** [comment]
Does it work?

**Comment by sri-harsh** [comment]
No, images are not being imported into the code I presume, that's why it is giving empty if I ain't wrong.

![image](https://user-images.githubusercontent.com/44131849/105405435-29e65780-5c51-11eb-8310-3fea3ecc526e.png)

**Comment by muhammedtalo** [comment]
Inside Train folder you sould have three folders (Covid/Pno/No_f)

**Comment by sri-harsh** [comment]
Yes, they are there, still the same problem, I have deleted all the files and re-uploaded and checked as well.

**Comment by muhammedtalo** [comment]
print train_fnames and let me see

**Comment by sri-harsh** [comment]
![image](https://user-images.githubusercontent.com/44131849/105406968-23f17600-5c53-11eb-9b56-6157095d7cc4.png)

Here...it is empty.

**Comment by muhammedtalo** [comment]
it should contain folders should not be emty.

**Comment by sri-harsh** [comment]
Yeah in the drive folders are there but it is showing empty.

**Comment by muhammedtalo** [comment]
For example, I have different problem and here is folders
![b](https://user-images.githubusercontent.com/26714468/105407430-c2bfa780-5c3e-11eb-8767-862d0f95a01c.JPG)

**Comment by muhammedtalo** [comment]
![e](https://user-images.githubusercontent.com/26714468/105407703-19c57c80-5c3f-11eb-8059-d098376454b3.JPG)

My drive structure for two classes

**Comment by muhammedtalo** [comment]
?

**Comment by sri-harsh** [comment]
The same problem persists to be there.

**Comment by muhammedtalo** [comment]
I hope you can solve it. Please let me know once you have it.

Best

**Comment by muhammedtalo** [comment]
I will provide a new repo for the above problem [Kidney stone detection] (hopefully) this week. You can also check that one.

**Comment by sri-harsh** [comment]
> In v2 colab
> 
> !pip install -Uqq fastbook
> import fastbook
> fastbook.setup_book()
> 
> from fastbook import *
> 
> from fastai.vision.all import *
> path = Path('/content/gdrive/My Drive/.../')
> train_fnames = get_image_files(path/"Train")
> 
> dblock = DataBlock(blocks=(ImageBlock(cls=PILImage), CategoryBlock),
> splitter=RandomSplitter(valid_pct=0.2, seed=23),
> get_y=parent_label,
> item_tfms=Resize(512, method="squish"),
> batch_tfms=None,
> )
> 
> dls = dblock.dataloaders(train_fnames)
> 
> train_df = len(train_fnames)
> print(f"There are {train_df} images")
> 
> # print class names, number of class, length of train dataset, and length of valid dataset
> dls.c, len(dls.train_ds), len(dls.valid_ds)
> 
> dls.show_batch()
> 
> ....

Hey worked till here, how do I proceed next?

**Comment by sri-harsh** [comment]
![image](https://user-images.githubusercontent.com/44131849/105411216-1f2fc080-5c59-11eb-9bde-f4b0a2ca7e3b.png)


Is it conv_block? Because there is no definition for conv_layer.

**Comment by muhammedtalo** [comment]
Check the new name for "conv_layer " in fastai -v2 from https://docs.fast.ai/
Please take a look at following tutorials: https://walkwithfastai.com/Introduction
Here is link how to apply 5-fold cross validation: https://walkwithfastai.com/Cross_Validation
Best,

**Comment by muhammedtalo** [comment]
The conv_layer in v1 have been named as conv2 in v2.
conv2(256,2)

**Comment by muhammedtalo** [comment]
Please let us know once you run the code.

**Comment by sri-harsh** [comment]
conv2(256,2) is not working for me.
ConvLayer(256, 2) is working.
But still facing the below issue.
![image](https://user-images.githubusercontent.com/44131849/105412634-045e4b80-5c5b-11eb-9edb-1ef5c150702c.png)

**Comment by muhammedtalo** [comment]
https://discuss.pytorch.org/t/runtimeerror-mat1-and-mat2-shapes-cannot-be-multiplied-64x13056-and-153600x2048/101315

**Comment by muhammedtalo** [comment]
If you have 3 folders your model output must be 3.

**Comment by muhammedtalo** [comment]
See: https://github.com/muhammedtalo/COVID-19/blob/master/DarkCovidNet%20model%20for%20three%20classes.ipynb

**Comment by sri-harsh** [comment]
> https://discuss.pytorch.org/t/runtimeerror-mat1-and-mat2-shapes-cannot-be-multiplied-64x13056-and-153600x2048/101315

Yeah changed linear related to filters to 1323.

**Comment by sri-harsh** [comment]
> If you have 3 folders your model output must be 3.

Got it, sorry got confused. Updated it accordingly.

max_lr is also not there in fastai v2, I removed max_lr as below, is it the right way to do it?

![image](https://user-images.githubusercontent.com/44131849/105414102-0c1eef80-5c5d-11eb-9b00-ee07e3110a32.png)

**Comment by muhammedtalo** [comment]
Perfect it is working now!. 
For different "lr "you can look at https://docs.fast.ai/callback.schedule.html#LRFinder and search the main website https://docs.fast.ai/
Good luck!

**Comment by sri-harsh** [comment]
The way you used max_lr
![image](https://user-images.githubusercontent.com/44131849/105415355-9e73c300-5c5e-11eb-9fbd-6e2f9f8a2475.png)

I removed max_lor
![image](https://user-images.githubusercontent.com/44131849/105415434-b64b4700-5c5e-11eb-8b23-2970e4e74aa3.png)

Is this fine? This is the first time I am using fast.ai, I googled regarding this, did not properly get it.

**Comment by muhammedtalo** [comment]
It is OK you can use without max. 
In the v1: https://forums.fast.ai/t/question-on-default-layer-learning-rates-and-fit-one-cycle/50950

**Comment by muhammedtalo** [comment]
This may help better 
https://towardsdatascience.com/finding-good-learning-rate-and-the-one-cycle-policy-7159fe1db5d6

**Comment by sri-harsh** [comment]
Thanks, for the reference links.

So, the only part left to be implemented in code is 5-fold cross validation right?

**Comment by muhammedtalo** [comment]
Yes, I have shared the link above for implementation (https://walkwithfastai.com/Cross_Validation)

**Comment by sri-harsh** [comment]
Hey, what did you use for training the model? I mean did you train it on google colab? for me each epoch is taking around 5-7 minutes whereas it hardly took 30 seconds for you as per https://github.com/muhammedtalo/COVID-19/blob/master/DarkCovidNet%20model%20for%20three%20classes.ipynb
